user/mgetty/contrib/ptylogin/ptylogin.1
.TH ptylogin 1 "10 january 1999"
.IX ptylogin
.SH NAME
ptylogin \-\ replacement for mgetty's login.config rlogin hack fixing
security and denial of service problems with ownership of tty.
.SH SYNOPSIS
.B ptylogin
login-name
.SH DESCRIPTION
This manual page documents
.BR ptylogin .
.B ptylogin
is launched from mgetty's login.config configuration file with root
priviledges. It opens a pty slave/master pair, and forks /bin/login.
It ensures data stream is 8 bit. This means that the user which
logs in will not be connected to the tty of the modem, but to
a pty. The pty slave will be owned (because of /bin/login) by the
logged-on user. The modem tty will be owned by root, and permissions
will be rw access for root only. That tty doesn't need, by the way, to
be logged-in. When the modem disconnects, the master pty is closed
and a SIGHUP is transmitted to the other side. The worse that the
user can then do is leave their process on if they disabled the
SIGHUP. However they can't access the modem device nor reopen it.

For enhanced security we assume the escape sequence of the modem is
disabled, and that a modem hangup from the user calling our local
modem causes a SIGHUP to the ptylogin process.

Please look at the
.B Paranoid Secure Port Implementation
RCS revision SPEC,v 1.6 1999/01/05 08:41:46 or
later for all details of the problem ptylogin fixes (it's
quite tricky).

.SS OPTIONS

.I "\login-name"
This must be a 8 char maximum login name to launch login into,
must exist, and may not contain \- or spaces. As
.B /bin/login
is not launched through system() but instead with exec(), common
attacks like semicolons or other separators cannot happen.

.SH EXAMPLES

The login.config could be configured like this:

.I "*       root  dialin  /usr/bin/ptylogin @"

Note that if you specify users which bypass this default, for
example for PPP, FTN or UUCP, you would enter something like

.I "uu*     \-       \-       /bin/login.one @"

.B WARNING:
You must use a login program which doesn't allow more than one retry. Else
interactive users can bypass the default ptylogin restricted login.

.SH AUTHOR
Marc SCHAEFER <schaefer@alphanet.ch>

.SH VERSION
Manual version 1.0 PV001 documents ptylogin version 1.0

.SH NOTES

.SH BUGS
Please look at the source.

.SH TODO

.SH BASED\-ON
\- An idea to simplify rlogin and still fix the problems from
     Theodore Y. Ts'o <tytso@MIT.EDU>

\- rlogind and rlogin code from Linux NetKit-0.09

\- virtual_connection from Marc SCHAEFER <schaefer@alphanet.ch>

.SH HISTORY

.SH COPYRIGHT
This work is (C) Marc SCHAEFER 1999 and has been done in my free
time. However, it is placed under the GPL and thus any use is
authorized as long as you do not prevent others from using it and
accessing the original source code or your extensions.

.SH DISCLAIMER
The author hereby disclaims any warranty, either expressed or
implied, regarding this software and documentation. The fact that
this software attempts to fix a security vulnerability doesn't mean
that it doesn't have any vulnerability, some which could be more
serious than the one it tries to fix.


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/mgetty/contrib/ptylogin/ptylogin.c
/* ptylogin.c -- links the current tty to a /bin/login process through
 *               a pty pair
 * Marc SCHAEFER <schaefer@alphanet.ch>
 * Creation: 10/01/98
 * Update:   10/01/98
 * V1.0 PV001
 * DESCRIPTION
 *    This emulates a rlogin -8E -l USER localhost from mgetty's login.conf
 *    without the need for two sockets, two processes, and networking
 *    running.
 *    This allows to prevent direct control of the modem tty by a
 *    malicious user, work-arounding many Denial of Service or even
 *    security problems. Please look at the ``Paranoid Secure Port
 *    Implementation'', RCS revision SPEC,v 1.6 1999/01/05 08:41:46 or
 *    later for all details.
 * USAGE
 *    See manpage.
 * RESULT
 *    0 everything fine
 *    1 couldn't exec() /bin/login or allocate pty.
 *    2 missing argument, illegal login name or not running as root.
 * NOTES
 *    - This should NOT be suid as it is not necessary and it has not
 *      been designed to be run outside of mgetty's login.config.
 * BUGS
 * TODO
 *   - Test
 *   - Do we need to ensure running as root for pty security ?
 *   - Do we need to change tty permission and owner or is it done by
 *     mgetty ?
 *   - Alternative pty allocation code not needing root ? (there is
 *     code floating around for Linux).
 *   - Ask people to review the code and the specification.
 * ASSUMPTIONS
 *    - This is called with root priviledges since this is more secure
 *      in term of tty ownership and pty allocation.
 * COPYING-POLICY
 *    (C) Marc SCHAEFER, under the GPL
 *    parts Copyright (c) 1983, 1988, 1989
 *             The Regents of the University of California
 * BASED-ON
 *    - A problem described by Marc SCHAEFER in the above Paranoid
 *      specification.
 *    - An idea to simplify rlogin and still fix the problems from
 *         "Theodore Y. Ts'o" <tytso@MIT.EDU>
 *    - rlogind and rlogin code from Linux NetKit-0.09, BSD license.
 *    - virtual_connection from Marc SCHAEFER <schaefer@alphanet.ch>, GPL.
 * MODIFICATION-HISTORY
 *    10/01/99  schaefer  Created this file.
 * $Id: ptylogin.c,v 1.1 1999/01/16 17:17:08 gert Exp $
 */

#include <stdio.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/errno.h>
#include <termios.h>
#include <sys/time.h>
#include <sys/ioctl.h>

/* Macro and definitions */

#define MAXIMUM_LOGIN_NAME_LENGTH 8
#define PTS_NAME_SIZE 20
#define _PATH_LOGIN "/bin/login"
#define BUFFER_LENGTH 4096

#undef  MGETTY_HAS_SET_OWNER_AND_MODE  /* ptylogin does it */

/* Public function definitions and hacks */

/* BUGS
 *    - Should be in libbsd.h or similar include file.
 */
pid_t forkpty(int *, char *, struct termios *, struct winsize *);

/* Private type definitions */

#ifndef fd_t
typedef int fd_t;
#endif /* !fd_t */

/* Private function definitions */

/* NAME
 *    connect_login
 * DESCRIPTION
 *    Connects this process to the /bin/login process through a pty
 *    pair. Do any local tty initialization (including chmod()/chown()).
 * RESULT
 *    exit code
 *       0 session terminated normally (hangup or exit)
 *       1 couldn't spawn child or open ptys.
 * NOTES
 * BUGS
 * TODO
 */
static int connect_login(char *login_name);

/* NAME
 *    reject_login_name
 * DESCRIPTION
 *    Verify the constraints on the login name, which are that it doesn't
 *    dashes (-) or spaces (only normal spaces), and it 8 characters
 *    long or less.
 * RESULT
 *    0 if the constraint DO apply and the login name should be
 *    accepted, else different of zero and the login should be rejected.
 * NOTES
 *    - We tested this function with 7, 8 and 9 characters. We also tested
 *      one hash at the end and beginning and one space at the end. In all
 *      case the function answered as expected.
 * BUGS
 * TODO
 *    - Theoretically those constraints are already mostly handled by
 *      mgetty. Check this. Until then we do it here.
 */
static int reject_login_name(char *login_name);

/* NAME
 *    set_raw
 * DESCRIPTION
 *    Sets the stdin tty to raw (if mode == 1), else cooked.
 * RESULT
 *    NONE
 * NOTES
 *    - Each call to set_raw(0) must have been after a corresponding
 *      and unnested call to set_raw(1) because of internal state.
 * BUGS
 * TODO
 */
static void set_raw(int mode);

/* Public constants */

char rcsid[] =
  "$Id: ptylogin.c,v 1.1 1999/01/16 17:17:08 gert Exp $";


/* Private constants */

/* Private variables */

static struct winsize win = { 0, 0, 0, 0 };
static struct termio saved; /* Internal state for set_raw() */

/* Function implementation */

int main(int argc, char **argv) {
   if (argc == 2) {
      if ((getuid() == 0) && (geteuid() == 0)) {
	 if (!reject_login_name(argv[1])) {
	    exit(connect_login(argv[1]));
	 }
	 else {
	    fprintf(stderr,
                    "%s: illegal length or character(s) in login name.\n", 
                    argv[0]);
	    exit(2);
	 }
         /* NOT REACHED */
      }
      else {
	 fprintf(stderr,
                 "%s: not running as root.\n",
                 argv[0]);
	 exit(2);         
      }
      /* NOT REACHED */
   }
   else {
      fprintf(stderr, "%s login-name\n%s: bad args.\n", argv[0], argv[0]);
      exit(2);
   }

   /* NOT REACHED */
}

static int reject_login_name(char *login_name) {
   return (strlen(login_name) > MAXIMUM_LOGIN_NAME_LENGTH)
          || (strchr(login_name, ' '))
          || (strchr(login_name, '-'));
}

static int connect_login(char *login_name) {
   pid_t pid;
   fd_t master_fd;
   char line[PTS_NAME_SIZE];

#ifndef MGETTY_HAS_SET_OWNER_AND_MODE
   if (fchown(0, 0, 0)) {
      perror("fchown()");
      return 1;
   }
   else if (fchmod(0, 0600)) {
      perror("fchmod()");
      return 1;
   }
#endif /* !MGETTY_HAS_SET_OWNER_AND_MODE */

   switch (pid = forkpty(&master_fd, line, NULL, &win)) {
      case 0:
         /* Child. Setup slave and exec /bin/login */

         /* ASSUMPTIONS
          *    - No other open file except slave pty, which is on 0/1/2.
          *    - Controlling tty set.
          */

         /* NOTES
          *    - Keeping whole environment.
          */
         execl(_PATH_LOGIN, "login", login_name, NULL);

         return 1; /* if arriving here, something wrong */
         break;
      default:
         /* Parent */
         if (pid < 0) {
            if (errno == ENOENT) {
               perror("Out of ptys");
            }
            else {
               perror("forkpty()");
            }
            return 1;
         }
         else {
            int ok = 1;
	    int nfound;
	    fd_set readfds;
	    fd_set writefds;
	    char from[BUFFER_LENGTH];
	    char to[BUFFER_LENGTH];
	    int infrom = 0;
	    int into = 0;
	    int temp;

            /* Transfer bytes transparently until SIGHUP or EOF */

            set_raw(1);

            while (ok) {
	       FD_ZERO(&readfds);
	       FD_ZERO(&writefds);

	       FD_SET(0, &readfds);
	       FD_SET(master_fd, &readfds);

	       if (infrom) {
		  FD_SET(1, &writefds);
	       }
	       if (into) {
		  FD_SET(master_fd, &writefds);
	       }

	       /* from: buffer used to store data from remote site to be output
		*       to stdout (1). (read from master_fds)
		* to:   buffer used to store data from `modem' to be output to
		*       socket. (read from 0).
		*/
	       nfound = select(master_fd + 1, &readfds, &writefds, NULL, NULL);
	       if (nfound) {
		  if (FD_ISSET(0, &readfds)) {
		     if (into < BUFFER_LENGTH) {
			temp = read(0, to + into, BUFFER_LENGTH - into);
			if (temp == -1 || temp == 0 ) {
			   ok = 0;
			}
			else {
			   into += temp;
			}
		     }
		     /* else we will do it later. */
		  }
		  if (FD_ISSET(master_fd, &readfds)) {
		     if (infrom < BUFFER_LENGTH) {
			temp = read(master_fd, from + infrom, BUFFER_LENGTH - infrom);
			if (temp == -1 || temp == 0) {
			   ok = 0;
			}
			else {
			   infrom += temp;
			}
		     }
		     /* else we will do it later. */
		  }
		  if (FD_ISSET(1, &writefds)) {
		     if (infrom > 0) {
			temp = write(1, from, infrom);
			if (temp == -1 ) {
			   ok = 0;
			}
			else {
			   infrom -= temp;
			}
		     }
		  }
		  if (FD_ISSET(master_fd, &writefds)) {
		     if (into > 0) {
			temp = write(master_fd, to, into);
			if (temp == -1 ) {
			   ok = 0;
			}
			else {
			   into -= temp;
			}
		     }
		  }
	       }
	    }

            set_raw(0);


            return 0;
         }
         break;
   }
   
   printf("connect");
   return 0;
}

void set_raw(int mode) {
   struct termio tty;

   /* ASSUMPTION
    *    - Doing it for 0 also does it for 1 (device-wide).
    */

   if (!mode) {
      ioctl(0, TCSETAW, &saved);
   }
   else {
      ioctl(0, TCGETA, &tty);
      ioctl(0, TCGETA, &saved);

      tty.c_iflag = (IGNBRK);

      /* No echo, crlf mapping, INTR, QUIT, delays, no erase/kill */
      tty.c_lflag &= ~(ECHO | ICANON | ISIG);

      tty.c_oflag = 0;        /* Transparent output */

      tty.c_cflag &= ~PARENB; /* Same baud rate, disable parity */
      tty.c_cflag |= CS8;     /* Set character size = 8 */

      /* NOTES
       *    - We assume mgetty set crtscts.
       */

      tty.c_cc[VMIN] = 1; /* This many chars satisfies reads */
      tty.c_cc[VTIME] = 1;    /* or in this many tenths of seconds */

      ioctl(0, TCSETAW, &tty);
   }
}
